composer/core/state.py

```diff
@@ -521,6 +608,15 @@ def __init__(
                 ),
             )
 
+        self.device_mesh: Optional[DeviceMesh] = _create_device_mesh(self.device, self.fsdp_config, self.tp_config)
+        if self.fsdp_config is not None and self.device_mesh is not None:
+            fsdp_mesh_dim_names = [ParallelismType.DATA_PARALLEL_SHARD.value]
+            if self.device_mesh.mesh_dim_names is not None and ParallelismType.DATA_PARALLEL_REPLICATE.value in self.device_mesh.mesh_dim_names:
+                fsdp_mesh_dim_names.append(ParallelismType.DATA_PARALLEL_REPLICATE.value)
+            self.fsdp_config['device_mesh'] = self.device_mesh[tuple(fsdp_mesh_dim_names)]  # type: ignore
```

Comment:
I actually think the order here might be not correct ? according to the pytorch doc:
1. 
<img width="887" alt="image" src="https://github.com/mosaicml/composer/assets/2545820/2f0c4c7e-325e-41a8-b232-612f3dfdfbc4">
2. [_inter_node_pg to be the outermost dimension, and the process_group(intra_node) to be the innermost dimension](https://github.com/pytorch/pytorch/blob/5325a6de64c9b11a8f35e59e709499135d23579c/torch/distributed/fsdp/_init_utils.py#L163-L166)

My understanding is that the outermost dimension (dim-0) should be replicate, and innermost dimension (dim-1) should be shard dimension. 

cc @snarayan21

Reply:
I think we might need 6 GPUs or 8 GPUs (4 X2 or 2 X 4) to test this , lol. 4 GPUs (2 X 2) can't test this order.

Reply:
Yep you're correct! Fixed

Reply:
It seems you only fixed the order when constructing the self.device_mesh [here in the beginning of this file](https://github.com/mosaicml/composer/pull/3269/commits/00ce8c654aa0d19a47b308210baa4e67a370e48d) , but here when you construct the device_mesh for fsdp, in `fsdp_mesh_dim_names`, the first one is still `DATA_PARALLEL_SHARD`, not `DATA_PARALLEL_REPLICATE` ?